compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVQueryBuilder.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVQueryBuilder.h"
+
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/Vulkan/IR/VulkanAttributes.h"
+#include "iree/compiler/Dialect/Vulkan/IR/VulkanDialect.h"
+#include "iree/compiler/Dialect/Vulkan/Utils/TargetEnvironment.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/SPIRV/IR/SPIRVAttributes.h"
+#include "mlir/Dialect/SPIRV/IR/SPIRVOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Value.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+// Return the subgroup feature bit according to the following table:
+//
+// VK_SUBGROUP_FEATURE_BASIC_BIT = 0x00000001
+// VK_SUBGROUP_FEATURE_VOTE_BIT = 0x00000002
+// VK_SUBGROUP_FEATURE_ARITHMETIC_BIT = 0x00000004
+// VK_SUBGROUP_FEATURE_BALLOT_BIT = 0x00000008
+// VK_SUBGROUP_FEATURE_SHUFFLE_BIT = 0x00000010
+// VK_SUBGROUP_FEATURE_SHUFFLE_RELATIVE_BIT = 0x00000020
+// VK_SUBGROUP_FEATURE_CLUSTERED_BIT = 0x00000040
+// VK_SUBGROUP_FEATURE_QUAD_BIT = 0x00000080
+// VK_SUBGROUP_FEATURE_PARTITIONED_BIT_NV = 0x00000100
+uint32_t processSubgroupCapability(spirv::Capability c) {
+  switch (c) {
+  case spirv::Capability::GroupNonUniform:
+    return 0x00000001;
+  case spirv::Capability::GroupNonUniformVote:
+    return 0x00000002;
+  case spirv::Capability::GroupNonUniformArithmetic:
+    return 0x00000004;
+  case spirv::Capability::GroupNonUniformBallot:
+    return 0x00000008;
+  case spirv::Capability::GroupNonUniformShuffle:
+    return 0x00000010;
+  case spirv::Capability::GroupNonUniformShuffleRelative:
+    return 0x00000020;
+  case spirv::Capability::GroupNonUniformClustered:
+    return 0x00000040;
+  case spirv::Capability::GroupNonUniformQuad:
+    return 0x00000080;
+  case spirv::Capability::GroupNonUniformPartitionedNV:
+    return 0x00000100;
+  default:
+    break;
+  }
+  return 0x00000000;
+}
+
+LogicalResult buildVulkanSPIRVDeviceRequirementQueries(
```

Comment:
This actually runs into some of the work I'm pushing on right now, see https://github.com/openxla/iree/pull/15888#issuecomment-1852515490. Sorry I think you had this stuff long in draft mode and what I pointed previously is pretty new. Let's chat a bit about how we should approach this and distribute the work. :)

Reply:
Sounds good, I was just pipe cleaning some PRs I had sitting around. Let's chat some time today.